### PR TITLE
Library for writing interactive "shell script" in Go

### PR DESCRIPTION
**Pull request by coryrc**
Advance knative/test-infra#2063

Already used elsewhere

/cc @chaodaiG 
/cc @chizhg

### CODE REVIEW

File: test/interactive/command.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// could be public so extensions of this can test themselves
+var defaultRun func(*exec.Cmd) error
+
+func init() {
+	defaultRun = func(c *exec.Cmd) error { return c.Run() }
+}
+
+// Command represents running an executable, to make it easy to write interactive "shell scripts" in Go.
+// This routes standard in, out, and error from and to the calling program, intended usually to be a login shell run by a person.
+type Command struct {
```
Comment (chizhg) [inline test/interactive/command.go:38]: I feel the functionality is a bit overlapping with https://github.com/knative/pkg/blob/master/test/cmd/command.go, any chance we can move it there?

File: test/interactive/docker.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	for k, v := range e {
+		d.AddArgs("-e", fmt.Sprintf("%s=%s", k, v))
+	}
+}
+
+// AddMount add arguments for the --mount command
+func (d *Docker) AddMount(typeStr, source, target string, optAdditionalArgs ...string) {
+	addl := ""
+	for _, arg := range optAdditionalArgs {
+		addl = addl + "," + arg
```
Comment (chizhg) [inline test/interactive/docker.go:73]: This will be `addl := strings.Join(arg, ",")`

File: test/interactive/docker.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		}
+	}
+	overlayDir := path.Join(tmpDir, "overlay")
+	// The options for overlay mount are confusing
+	// You need empty directories for upper and work (and overlay, though you can mount over directories that have files in them if you *want* to...)
+	mount := NewCommand("sudo", "mount", "-t", "overlay", "-o",
+		fmt.Sprintf("lowerdir=%s,upperdir=%s/upper,workdir=%s/work", externalDirectory, tmpDir, tmpDir),
+		"none", overlayDir)
+	// Print command to run so user knows why it is asking for sudo password (if it does)
+	fmt.Println(mount)
```
Comment (chizhg) [inline test/interactive/docker.go:102]: Use `log` and be a little bit verbose?

File: test/interactive/docker.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	if err = mount.Run(); err != nil {
+		log.Fatalf("Unable to create overlay mount, so giving up: %v", err)
+	}
+	d.AddMount("bind", overlayDir, internalDirectory)
+	return func() {
+		// Print command to run so user knows why it is asking for sudo password (if it does)
+		umount := NewCommand("sudo", "umount", overlayDir)
+		fmt.Println(umount)
```
Comment (chizhg) [inline test/interactive/docker.go:110]: Ditto?

File: test/interactive/docker_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	"os/exec"
+	"testing"
+)
+
+func TestEnv(t *testing.T) {
+	e := make(Env)
+	err := e.PromoteFromEnv("PWD")
+	if err != nil {
+		t.Error(err)
+		fmt.Printf("The Environ is:\n%+v\n", os.Environ())
```
Comment (chizhg) [inline test/interactive/docker_test.go:31]: `t.Logf`

File: test/interactive/docker_test.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	if e["PWD"] != os.Getenv("PWD") {
+		t.Errorf("Promotion did not occur correctly: %v;%v", e["PWD"], os.Getenv("PWD"))
```
Comment (chizhg) [inline test/interactive/docker_test.go:34]:
```suggestion
		t.Errorf("Promotion for $PWD did not occur correctly: %s != %s", e["PWD"], os.Getenv("PWD"))
```

File: test/interactive/docker_test.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	badName := "GEEZ_I_REALLY_H0PE_TH1S_DOES_NOT_EXIST"
```
Comment (chizhg) [inline test/interactive/docker_test.go:36]: 👍

File: test/interactive/command.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	Name string
+	Args []string
+	// If LogFile is set, anything send to stdout and stderr is tee'd to the file named in LogFile when .Run() is called.
+	LogFile string
```
[the remark this answers is not in the document]
Comment (coryrc) [reply test/interactive/command.go:42]: if `c := NewCommand()`, then `c.LogFile = "..."` is the way I use it

File: test/interactive/docker.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// PromoteFromEnv pulls the named environment variables from the environment and puts them in the Env.
+// It does not stop on error and returns an error listing all the failed values
+func (e Env) PromoteFromEnv(envVars ...string) error {
+	var err error
+	for _, env := range envVars {
+		v := os.Getenv(env)
+		if v == "" {
+			err = fmt.Errorf("Environment Variables '%s' is not set; %v", env, err)
```
[the remark this answers is not in the document]
Comment (coryrc) [reply test/interactive/docker.go:49]: Huh, will change. I like it including all.

File: test/interactive/command.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (coryrc) [reply test/interactive/command.go:42]: but could use a Setter instead if you want

File: test/interactive/command.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+// String conforms to Stringer
+func (c Command) String() string {
+	s := c.Name + " " + strings.Join(c.Args, " ")
+	return fmt.Sprintf("Command to run: %s", s)
+}
+
+// AddArgs appends arguments to the current list of args
+func (c *Command) AddArgs(args ...string) {
+	for _, a := range args {
```
[the remark this answers is not in the document]
Comment (coryrc) [reply test/interactive/command.go:81]: d'oh, I always forget that

File: test/interactive/command.go (hunk 1 above)
Comment (coryrc) [reply test/interactive/command.go:38]: That library explicitly runs commands in a different environment than the caller's environment, while this library is intended for commands running in the same environment (like a shell script). Given it's just a tiny wrapper around exec.Command(), there's nothing else to combine.

File: test/interactive/docker.go (hunk 2 above)
Comment (coryrc) [reply test/interactive/docker.go:73]: the leading `,` is necessary with arguments, which rules out your first example, or necessary that the string is empty when no arguments, ruling out the second example unless we add `if len(optAdditionaArgs) > 0` first. Your call

File: test/interactive/docker_test.go (hunk 6 above)
Comment (coryrc) [reply test/interactive/docker_test.go:34]: actually, e["PWD"] is not guaranteed either, so will make a couple changes

File: test/interactive/docker_test.go (hunk 5 above)
Comment (coryrc) [reply test/interactive/docker_test.go:31]: oops

File: test/interactive/docker.go (hunk 3 above)
Comment (coryrc) [reply test/interactive/docker.go:102]: I use log elsewhere so should be consistent.

File: test/interactive/command.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply test/interactive/command.go:42]: I would prefer two functions - `NewCommand` and `NewCommandWithLogFile`, but up to you.

File: test/interactive/docker.go (hunk 2 above)
Comment (chizhg) [reply test/interactive/docker.go:73]: I think below is more understandable.
```
if len(optAdditionaArgs) != 0 {
    addl = "," + strings.Join(optAdditionalArgs, ",")
}
```

File: test/interactive/command.go (hunk 1 above)
Comment (chizhg) [reply test/interactive/command.go:38]: SG

File: test/interactive/docker.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+func init() {
+	defaultDockerCommands = []string{"docker", "run", "-it", "--rm", "--entrypoint", "bash"}
+}
+
+// Env represents a collection of environment variables and their values
+type Env map[string]string
+
+// Docker is mostly an Command preloaded with arguments which setup Docker for running an image interactively.
+type Docker struct {
+	Command
```
Comment (chizhg) [inline test/interactive/docker.go:39]: Nit: should `Command` be private so change to `command`?

Comment (coryrc) [reply test/interactive/docker.go:39]: then .Run and .AddArgs aren't available

File: test/interactive/command.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+	}
+	return c.run(cmd)
+}
+
+// String conforms to Stringer
+func (c Command) String() string {
+	s := c.Name + " " + strings.Join(c.Args, " ")
+	return fmt.Sprintf("Command to run: %s", s)
```
Comment (vagababov) [inline test/interactive/command.go:76]:
```suggestion
	"Command to run: " + s
```

File: test/interactive/docker.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// PromoteFromEnv pulls the named environment variables from the environment and puts them in the Env.
+// It does not stop on error and returns an error listing all the failed values
+func (e Env) PromoteFromEnv(envVars ...string) error {
+	var err error
+	for _, env := range envVars {
+		v := os.Getenv(env)
+		if v == "" {
+			err = fmt.Errorf("environment variable '%s' is not set; %v", env, err)
```
Comment (vagababov) [inline test/interactive/docker.go:49]:
```suggestion
			err = fmt.Errorf("environment variable %q is not set; %w", env, err)
```

File: test/interactive/command.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// Command represents running an executable, to make it easy to write interactive "shell scripts" in Go.
+// This routes standard in, out, and error from and to the calling program, intended usually to be a login shell run by a person.
+type Command struct {
+	Name string
+	Args []string
+	// If LogFile is set, anything send to stdout and stderr is tee'd to the file named in LogFile when .Run() is called.
+	LogFile string
+	run     func(*exec.Cmd) error
```
[the remark this answers is not in the document]
Comment (coryrc) [reply test/interactive/command.go:43]: If we were to have a complicated function like AddRWOverlay() with nested calls to NewCommand() that we could actually test in a useful way, defaultRun allows you to test those nested calls without actually running commands.

File: test/interactive/docker.go (hunk 13 above)
Comment (coryrc) [reply test/interactive/docker.go:49]: I don't wish to support unwrap so I'll just take the first

File: test/interactive/command.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (coryrc) [reply test/interactive/command.go:42]: The second leads to having lots of combinations; I could do the "Option" like the runcommands library

File: test/interactive/command.go (hunk 12 above)
Comment (coryrc) [reply test/interactive/command.go:76]: done

**Pull request merged**